PATA support

brynolv opened the pull request:
This is mega-scuffed. I welcome ideas to get rid of the `print!`s.

Tested in both debug and release, R-W x2 and W-R x2.

brynolv committed:
Started working on PATA

Added ports and read_sectors (still untested)

brynolv committed:
R/W implemented

Tested to work. However, a better method of delay than
print! would be prefered.

brynolv committed:
Debugging progressed

It works... on occasion. I I'm not sure why.

brynolv committed:
Functional

I do not know why the print!s make it work, but without them
the disk times out... At least it looks kinda neat.

Tested on debug and release.

Elekrisk commented on src/harddisk/pata.rs:
```diff
@@ -1,8 +1,32 @@
 
+use x86_64::instructions::port::{Port, PortReadOnly, PortWriteOnly};
+
+use crate::{nop, svec::SVec};
```
`crate::nop` doesn't exist in `crate`

brynolv replied:
Not *anymore*... and never in this repo. I'll run `rustfmt`as well, while I'm at it.

Elekrisk reviewed (changes requested):
Only one thing pops out to(?) me.

brynolv committed:
Removed local unused import and ran fmt

crate::nop only ever existed locally, and was ultimetely not used.

brynolv committed:
Greatly extended init checks

Does the bus exist, does it have a drive, is it ATA, does it support LBA48

brynolv committed:
Fixed time-out when doing immediate W/R

TL;DR: Forgot to check that cache flush was finished.

This meant that the read command was rejected, and with poll()
expecting a data-transfer, it spun until time-out.

Also made the fact that it's a time-out clearer in panic text.

Vikdos reviewed (approved):
ok:tm:

Elekrisk commented on src/harddisk/pata.rs:
```diff
@@ -1,37 +1,229 @@
+        }
+        loop {
+            let status = STATUS_REG.read();
+            if status & 8 == 0 {
```
Shouldn't this be `0x80`? according to oswiki

Elekrisk replied:
Actually, no, i read wrong :P

Elekrisk replied:
I did read wrong, but there is still something wrong.
It should probs be `== 8`

Elekrisk reviewed (changes requested):
Noticed a thing :P

Elekrisk reviewed (approved):
nvm lol

Elekrisk reviewed (changes requested):
There was something wrong lol

brynolv commented:
Draft till LBA28 (what's in QEMU) is supported

brynolv committed:
Added LBA28 support

Not yet tested.

brynolv commented:
It works as well as it did before

brynolv committed:
Added CPU-speed compensation

Probably slower than it *needs* to be, but if it works, it works!

brynolv committed:
Tweaked MAX_ITER

Like moving it to the correct place... Oops.
